### ISSUE TITLE
Unexpected behavior when subscribing to derived atoms in a specific order

### ISSUE THREAD

**Issue by awaw00** [pmndrs/jotai#1956]
I've encountered a bug in Jotai when subscribing to derived atoms in a specific order. The issue arises when a derived atom's response function doesn't execute as expected due to the order of atom subscription.

Here's a code snippet to reproduce the issue:

```jsx
import { createStore, atom, useAtomValue, useSetAtom } from 'jotai';

const valueAtom = atom(0);
valueAtom.debugLabel = 'value'

const dep1Atom = atom((get) => get(valueAtom) * 2);
dep1Atom.debugLabel = 'dep1'

const dep2Atom = atom((get) => {
  return `${get(valueAtom)}/${get(dep1Atom)}`;
});
dep2Atom.debugLabel = 'dep2'

const dep3Atom = atom((get) => {
  return `${get(dep1Atom)}`;
})
dep3Atom.debugLabel = 'dep3'

const store = createStore();

const logAtom = (a) => {
  const val = store.get(a);

  const line = document.createElement('div');
  line.innerText = `${a.debugLabel}: ${val}`;
  document.body.appendChild(line);
}

store.sub(dep2Atom, () => logAtom(dep2Atom));
store.sub(dep1Atom, () => logAtom(dep1Atom));
store.sub(valueAtom, () => logAtom(valueAtom));
store.sub(dep3Atom, () => logAtom(dep3Atom));

const btn = document.createElement('button');
btn.innerText = 'add';
btn.onclick = () => store.set(valueAtom, store.get(valueAtom) + 1);
document.body.appendChild(btn);
```

In the above code, atoms are subscribed in the order of dep2, dep1, value, and dep3. When a new value is written to valueAtom, the response function of dep3 is not executed. This is because during the process of handling the write to valueAtom, dep2 is processed first due to the order of subscription. While processing dep2, the value of dep1 (a dependency of dep2) is also updated via readAtomState, causing the value of dep2 to be updated prematurely.

Because the recomputeDependents method for dep1 is not executed until after this update, the comparison between prevAtomState and nextAtomState for dep1 fails, leading to dep3 (which only depends on dep1) not being updated.

This behavior seems to be inconsistent with the expected behavior. I would appreciate it if you could look into this issue. Thank you.

----
Reproduction with useAtomValue can see https://codesandbox.io/s/throbbing-fire-5fq563?file=/src/App.tsx

**Comment by dai-shi** [comment]
Hi, thanks for reporting.
Can you confirm if the issue is reproducible only with store api?
Or, can you reproduce the issue with `useAtom` without `createStrore`?

**Comment by awaw00** [comment]
> Hi, thanks for reporting. Can you confirm if the issue is reproducible only with store api? Or, can you reproduce the issue with `useAtom` without `createStrore`?

Yes, this issue can be reproduced using createStore, useAtom, or useValueAtom, provided that atoms are subscribed in a specific order (dep2 -> dep1 -> value -> dep3). Here is a code snippet that reproduces the issue:

```jsx
import React from 'react';
import ReactDOM from 'react-dom';
import { createStore, atom, useAtomValue, useSetAtom } from 'jotai';

const valueAtom = atom(0);
valueAtom.debugLabel = 'value'

const dep1Atom = atom((get) => get(valueAtom) * 2);
dep1Atom.debugLabel = 'dep1'

const dep2Atom = atom((get) => {
  return `${get(valueAtom)}/${get(dep1Atom)}`;
});
dep2Atom.debugLabel = 'dep2'

const dep3Atom = atom((get) => {
  return `${get(dep1Atom)}`;
})
dep3Atom.debugLabel = 'dep3'

const Child3 = () => {
  const dep2 = useAtomValue(dep2Atom);
  const setValue = useSetAtom(valueAtom);

  return (
    <>
      <div>dep2: {dep2}</div>
      <div>
        <button onClick={() => setValue((v) => v + 1)}>add</button>
      </div>
    </>
  );
}

const Child2 = () => {
  const dep1 = useAtomValue(dep1Atom);

  return (
    <>
      <div>dep1: {dep1}</div>
      <Child3 />
    </>
  );
}

const Child1 = () => {
  const value = useAtomValue(valueAtom);

  return (
    <>
      <div>value: {value}</div>
      <Child2 />
    </>
  );
}

const Root = () => {
  const dep3 = useAtomValue(dep3Atom);
  return (
    <>
      <div>dep3: {dep3}</div>
      <Child1 />
    </>
  );
}

// const store = createStore();

// const logAtom = (a) => {
//   const val = store.get(a);

//   const line = document.createElement('div');
//   line.innerText = `${a.debugLabel}: ${val}`;
//   document.body.appendChild(line);
// }

// store.sub(dep2Atom, () => logAtom(dep2Atom));
// store.sub(dep1Atom, () => logAtom(dep1Atom));
// store.sub(valueAtom, () => logAtom(valueAtom));
// store.sub(dep3Atom, () => logAtom(dep3Atom));

// const btn = document.createElement('button');
// btn.innerText = 'add';
// btn.onclick = () => store.set(valueAtom, store.get(valueAtom) + 1);
// document.body.appendChild(btn);

ReactDOM.render(<Root />, document.getElementById('mountNode'))

```

**Comment by dai-shi** [comment]
Okay, even without `createStore`. That's bad.

Can you please create a reproduction with https://csb.jotai.org, and put the link in the description of this issue?

Just curious, you have 4 atoms, can't it be reproduced with 3 atoms?

**Comment by awaw00** [comment]
The reproduction link (https://codesandbox.io/s/throbbing-fire-5fq563?file=/src/App.tsx) has been updated in the issue description. From what I've observed, 4 atoms are necessary, but the subscription to valueAtom and dep1Atom can be omitted.

Here is the implementation of recomputeDependents:

```js
  var recomputeDependents = function recomputeDependents(atom) {
    var mounted = mountedMap.get(atom);
    mounted == null ? void 0 : mounted.t.forEach(function (dependent) {
      if (dependent !== atom) {
        var prevAtomState = getAtomState(dependent);
        var nextAtomState = readAtomState(dependent);

        if (!prevAtomState || !isEqualAtomValue(prevAtomState, nextAtomState)) {
          recomputeDependents(dependent);
        }
      }
    });
  };
```

Let's analyze it in the context of our scenario.

When we assign a value to `valueAtom`, the following processes occur:

1. `recomputeDependents(valueAtom)`
2. `mountedMap.get(valueAtom).t.forEach(loop)`
3. Due to the dependency relationship and registration order of the 4 atoms, the value of `mountedMap.get(valueAtom).t` is `[dep2Atom, valueAtom, dep1Atom]`.
4. `dep2Atom` enters the loop first, and the process of `readAtomState(dep2Atom)` updates the state of `dep1Atom`.
5. When `dep1Atom` enters the loop, it cannot correctly call `recomputeDependents(dep1Atom)` because its state has already been updated.

Finally, since `dep3Atom` only depends on `dep1Atom`, we find that the update of `dep3Atom` is skipped.

**Comment by dai-shi** [comment]
Thanks for the reproduction. I added a test case in #1959.

**Comment by dai-shi** [comment]
btw, it works in v1: https://codesandbox.io/s/determined-volhard-fuzgf0
so, this is a regression in v2.

**Comment by dai-shi** [comment]
This is already fixed in #1959.
Please try the latest version (v2.4.2 as of now).
